python/pyarrow/dataset.py

```diff
@@ -153,15 +153,14 @@ def partitioning(schema=None, field_names=None, flavor=None):
 
 def _ensure_fs(filesystem, path):
     # Validate or infer the filesystem from the path
-    from pyarrow.fs import FileSystem, LocalFileSystem
+    from pyarrow.fs import FileSystem, LocalFileSystem, FileType
 
     if filesystem is None:
-        try:
+        # first check if the file exists as a local (relative) file path
+        filesystem = LocalFileSystem()
+        infos = filesystem.get_target_infos([path])[0]
+        if infos == FileType.NonExistent:
```

Comment:
Hmm, there's something fishy here. Shouldn't it be `infos.type == ...`?

Reply:
Sorry, this is a typo, that should certainly be `infos.type`. 
(and all tests are passing because we don't yet have any test that passes an URI that is not for a local file system ..)

Reply:
> (and all tests are passing because we don't yet have any test that passes an URI that is not for a local file system ..)

Hmm, I thought so. Then you should certainly add some tests.

Reply:
> Edit: hmm, that may not be that easy. I thought of using mockfs but for now it will probably create a new instance every time you pass a URI. Perhaps we need a way to create named mockfs instances.

Yeah, I first though to use mockfs as well. But indeed, that's not possible right now to point to an existing mockfs with an URI.

But added a test with S3 then. I confirmed that it actually failed with the original typo.